// @guildedts/rest/dist/GuildedAPIError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildedAPIError = void 0;
/**
 * Represents an error that occurred while interacting with the Guilded API.
 * @example new GuildedAPIError('ErrorCode', 'Message', 404, 'GET', '/channels/abc);
 */
class GuildedAPIError extends Error {
    code;
    message;
    status;
    method;
    path;
    body;
    meta;
    /**
     * @param code The code of the error.
     * @param message The message of the error.
     * @param status The status code of the error.
     * @param method The method used to make the request.
     * @param path The path used to make the request.
     * @param body The body used to make the request.
     * @param meta The meta data of the error.
     */
    constructor(code, message, status, method, path, body, meta) {
        super(message);
        this.code = code;
        this.message = message;
        this.status = status;
        this.method = method;
        this.path = path;
        this.body = body;
        this.meta = meta;
        this.name = `GuildedAPIError [${code}:${status}]`;
    }
}
exports.GuildedAPIError = GuildedAPIError;
