// @guildedts/rest/dist/routers/Router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const CalendarEventRouter_1 = require("./calendarEvent/CalendarEventRouter");
const CalendarEventRsvpRouter_1 = require("./calendarEvent/CalendarEventRsvpRouter");
const ServerBanRouter_1 = require("./server/ServerBanRouter");
const ServerMemberRouter_1 = require("./server/ServerMemberRouter");
const ServerRoleRouter_1 = require("./server/ServerRoleRouter");
const ServerRouter_1 = require("./server/ServerRouter");
const BaseRouter_1 = require("./BaseRouter");
const ChannelRouter_1 = require("./ChannelRouter");
const DocRouter_1 = require("./DocRouter");
const ForumTopicRouter_1 = require("./ForumTopicRouter");
const GroupRouter_1 = require("./GroupRouter");
const ListItemRouter_1 = require("./ListItemRouter");
const MessageRouter_1 = require("./MessageRouter");
const ReactionRouter_1 = require("./ReactionRouter");
const WebhookRouter_1 = require("./WebhookRouter");
/**
 * The router for the Guilded REST API.
 * @example new Router(rest);
 */
class Router extends BaseRouter_1.BaseRouter {
    /** The channel router for the Guilded REST API. */
    channels;
    /** The message router for the Guilded REST API. */
    messages;
    /** The server member router for the Guilded REST API. */
    serverMembers;
    /** The server ban router for the Guilded REST API. */
    serverBans;
    /** The forum topic router for the Guilded REST API. */
    forumTopics;
    /** The list item router for the Guilded REST API. */
    listItems;
    /** The doc router for the Guilded REST API. */
    docs;
    /** The reaction router for the Guilded REST API. */
    reactions;
    /** The server role router for the Guilded REST API. */
    serverRoles;
    /** The group router for the Guilded REST API. */
    groups;
    /** The webhook router for the Guilded REST API. */
    webhooks;
    /** The server router for the Guilded REST API. */
    servers;
    /** The calendar event router for the Guilded REST API. */
    calendarEvents;
    /** The calendar event RSVP router for the Guilded REST API. */
    calendarEventRsvps;
    /** @param rest The REST API manager that owns this router. */
    constructor(rest) {
        super(rest);
        this.channels = new ChannelRouter_1.ChannelRouter(rest);
        this.messages = new MessageRouter_1.MessageRouter(rest);
        this.serverMembers = new ServerMemberRouter_1.ServerMemberRouter(rest);
        this.serverBans = new ServerBanRouter_1.ServerBanRouter(rest);
        this.forumTopics = new ForumTopicRouter_1.ForumTopicRouter(rest);
        this.listItems = new ListItemRouter_1.ListItemRouter(rest);
        this.docs = new DocRouter_1.DocRouter(rest);
        this.reactions = new ReactionRouter_1.ReactionRouter(rest);
        this.serverRoles = new ServerRoleRouter_1.ServerRoleRouter(rest);
        this.groups = new GroupRouter_1.GroupRouter(rest);
        this.webhooks = new WebhookRouter_1.WebhookRouter(rest);
        this.servers = new ServerRouter_1.ServerRouter(rest);
        this.calendarEvents = new CalendarEventRouter_1.CalendarEventRouter(rest);
        this.calendarEventRsvps = new CalendarEventRsvpRouter_1.CalendarEventRsvpRouter(rest);
    }
}
exports.Router = Router;
